Update Docs for v5 changes. (#636)
reduxjs/reselect: README.md

# Reselect

[![npm package][npm-badge]][npm][![Coveralls][coveralls-badge]][coveralls][![GitHub Workflow Status][build-badge]][build]![TypeScript][typescript-badge]

A library for creating memoized "selector" functions. Commonly used with Redux, but usable with any plain JS immutable data as well.

> - Selectors can compute derived data, allowing `Redux` to store the minimal possible state.

Reviewer: :hand: style note: we typically leave the name of the library as a proper capitalized name ("Reselect", "Redux", "React-Redux" "Redux Toolkit"), and specifically use backticks for the lower-case actual package name (`reselect`, `redux`, `react-redux`, `@reduxjs/toolkit`).  Could you fix the names?

Author: Yes sir, sorry about that. I will fix that right away!

- Selectors are efficient. A selector is not recomputed unless one of its arguments changes.
- Selectors are composable. They can be used as input to other selectors.

The **Redux docs usage page on [Deriving Data with Selectors](https://redux.js.org/usage/deriving-data-selectors)** covers the purpose and motivation for selectors, why memoized selectors are useful, typical `Reselect` usage patterns, and using selectors with `React-Redux`.

## Installation

### Redux Toolkit

While `Reselect` is not exclusive to `Redux`, it is already included by default in [the official Redux Toolkit package](https://redux-toolkit.js.org) - no further installation needed.

```ts
import { createSelector } from '@reduxjs/toolkit'
```

### Standalone

For standalone usage, install the `Reselect` package:

> #### Using `npm`

Reviewer: :hand: Two notes:

- Not sure we to list all four package managers separately
- However many we leave in, consolidate them into a single code block ( `bash` language, and a `# NAME` comment above, like:

```bash
# NPM
npm install reselect

# Yarn
yarn add reselect
```

Author: I only did that to make copy pasting with a single click easier. Do you still want me to change it?

Reviewer: I don't think we need to optimize for people copy-pasting the instructions to install one package, yeah :)  We also use the code block approach in the other Redux docs sites for this step anyway.

Author: ok :(

```bash
npm install reselect
```

#### Using `yarn`

```bash
yarn add reselect
```

#### Using `bun`

```bash
bun add reselect
```

#### Using `pnpm`

```bash
pnpm add reselect
```

---

## Basic Usage

> `Reselect` exports a [`createSelector`] API, which generates memoized selector functions. [`createSelector`] accepts one or more [`input selectors`], which extract values from arguments, and a [`combiner`] function that receives the extracted values and should return a derived value. If the generated [`output selector`] is called multiple times, the output will only be recalculated when the extracted values have changed.

Reviewer: :hand: 

- There's a bunch of inline code names that also have brackets.  Are these supposed to be links?
- Typically inline code formatting is used for actual variable / function names.  "combiner" and "input selector" are descriptive terms and shouldn't be formatted as code.

Author: - Yes the brackets are links.
- I will fix the inline formatting.

Thanks for pointing these out!

}

const state: RootState = {
  todos: [
    { id: 0, completed: false },
    { id: 1, completed: true }
  ],
  alerts: [
    { id: 0, read: false },
    { id: 1, read: true }
  ]
}

const selectCompletedTodos = (state: RootState) => {
  console.log('selector ran')
  return state.todos.filter(todo => todo.completed === true)
}

selectCompletedTodos(state) // selector ran
selectCompletedTodos(state) // selector ran
selectCompletedTodos(state) // selector ran

const memoizedSelectCompletedTodos = createSelector(
  [(state: RootState) => state.todos],
  todos => {
    console.log('memoized selector ran')
    return todos.filter(todo => todo.completed === true)
  }
)

memoizedSelectCompletedTodos(state) // memoized selector ran
memoizedSelectCompletedTodos(state)
memoizedSelectCompletedTodos(state)

console.log(selectCompletedTodos(state) === selectCompletedTodos(state)) //=> false

console.log(
  memoizedSelectCompletedTodos(state) === memoizedSelectCompletedTodos(state)
) //=> true
```

As you can see from the example above, `memoizedSelectCompletedTodos` does not run the second or third time, but we still get the same return value as last time.

> Another difference is that with `memoizedSelectCompletedTodos` the referential integrity of the return value is also maintained through multiple calls of the selector, but the same cannot be said about the first example.

Reviewer: :hand: the phrasing of this sentence feels a bit awkward.  How about:

```suggestion
In addition to skipping unnecessary recalculations, `memoizedSelectCompletedTodos` returns the existing result reference if there is no recalculation.  This is important for libraries like React-Redux or React that often rely on reference equality checks to optimize UI updates.
```

  - [The default memoization function is no good, can I use a different one?](#the-default-memoization-function-is-no-good-can-i-use-a-different-one)
  - [How do I test a selector?](#how-do-i-test-a-selector)
  - [Can I share a selector across multiple component instances?](#can-i-share-a-selector-across-multiple-component-instances)
  - [Are there TypeScript Typings?](#are-there-typescript-typings)
  - [I am seeing a TypeScript error: `Type instantiation is excessively deep and possibly infinite`](#i-am-seeing-a-typescript-error-type-instantiation-is-excessively-deep-and-possibly-infinite)
  - [How can I make a curried selector?](#how-can-i-make-a-curried-selector)
  - [How can I make pre-typed version of `createSelector` for my root state?](#how-can-i-make-pre-typed-version-of-createselector-for-my-root-state)
  - [What if I want to use `createSelector` without memoization?](#what-if-i-want-to-use-createselector-without-memoization)
- [External References](#external-references)
- [Related Projects](#related-projects)
- [License](#license)
- [Prior Art and Inspiration](#prior-art-and-inspiration)

---

## Terminology

<details><summary><b>Click to expand</b></summary>

> - <a name="selector-function"></a>[**`Selector Function`**](#selector-function): A function that takes the `Redux` store state (or a part of it) as an argument and returns data derived from that state.

Reviewer: :hand: I think it's worth keeping the "not actually required to use Redux" aspect here.  How about:

```suggestion
- <a name="selector-function"></a>[**`Selector Function`**](#selector-function): A function that accepts one or more JS values as input arguments, and derives a result.  When used with Redux, the first argument is typically the entire Redux store state.
```

> - <a name="input-selectors"></a>[**`input selectors`**](#input-selectors): Basic selector functions used as building blocks for creating a memoized selector. They are passed as the first argument(s) to [`createSelector`].

Reviewer: ```suggestion
- <a name="input-selectors"></a>[**`input selectors`**](#input-selectors): Basic selector functions used as building blocks for creating a memoized selector. They are passed as the first argument(s) to [`createSelector`], are called with all of the selector arguments, and extract values for use in the result function.
```

- <a name="output-selector"></a>[**`Output Selector`**](#output-selector): The actual memoized selectors created by [`createSelector`].
- <a name="result-function"></a>[**`Result Function (Combiner)`**](#result-function): The function that comes after the [`input selectors`]. It takes the [`input selectors`]' return values as arguments and returns a result. Otherwise known as the `combiner`.

> - <a name="combiner"></a>[**`Combiner`**](#combiner): A function that takes [`input selectors`]' return values as arguments and returns a result (This function comes after the [`input selectors`] inside [`createSelector`]). This term is somewhat interchangeably used with `Result Function` or `resultFunc`. But `combiner` is more of a general term and `resultFunc` is more specific to `Reselect`. So the `resultFunc` is a `combiner` but a `combiner` is not necessarily the same as `resultFunc`. For the sake of simplicity, they are used synonymously throughout this documentation.

Reviewer: :hand: if we're going to consolidate terms, I'd like to drop "combiner" entirely and just stick with "result function" (or possibly "output function").

Author: Sounds good, the only reason I included "combiner" is because of how it was referened internally. But you're right we can drop that term in the docs to avoid further confusion. I think we can just stick to "result function". "output function" can be confused with "output selector". What do you think?

Reviewer: Oh. I missed the use of "output selector" here.  I don't think I like that usage tbh :)  I personally have used that phrase to mean what we're now calling the "result function".  If we have to distinguish things, I'd say "generated selector", maybe?

Author: Honestly I think "output selector" makes sense for the generated selector. It would also be consistent with how the types were named initially. The result function is not really a "selector" if you think about it. The way it is now we have "input selectors", then "result function" does the calculation, and we get an "output selector". What do you think?

Author: So it's kind of like:

input -> calculation -> output

Reviewer: eh, I'm also not sold on the type names either, even though they've been there since early on :)  "Parametric selector" has always confused me.

I think I'd like to stick with either "generated selector" or "memoized selector" as the term for the function returned by `createSelector`.

Author: > eh, I'm also not sold on the type names either, even though they've been there since early on :) "Parametric selector" has always confused me

Tbh, me neither, but I was just trying to stay consistent with the type names. And you're right about the "parametric selector", I'm not really a fan of it either. But they were there since the beginning and I thought that's how you would want me to name things in the docs as well.

Author: > I think I'd like to stick with either "generated selector" or "memoized selector" as the term for the function returned by `createSelector`.

"memoized selector" makes sense. However, since we have "input selectors" and if we have an "input" it makes sense that we would have an "output" as well. So I still think "output selector" makes the most sense.

Author: Also we have 2 types that I would like to remove entirely if that's okay with you. `ParametricSelector` and `OutputParametricSelector` since `Selector` and `OutputSelector` support additional parameters.

Reviewer: Yeah, go ahead and remove those. Breaking change, we'll have to document that, but might as well.

Per the "input/output" thing: I think I disagree because the "input selectors" are "selectors" (args in, derived value out even if it's usually just an extraction), and "inputs to `createSelector`".  The "result function" doesn't get any of the original args, so even though I've _referred_ to it as an "output selector" in the past, that's really not a good usage and I shouldn't have called it that.

The actual generated selector isn't just about "output", though.  It has all three pieces: the input arguments, the memoization, and calculating the final result.  It's not _just_ about "output".  The overall signature is `(...args) => derivedValue`, so it's a standard selector overall, it just happens to be a "memoized" selector that was "generated" by `createSelector` rather than hand-written.  So, I do think that "memoized" or "generated" are the best terms here.

Author: Yeah the "result function" is not an "output selector". I agree. That was one of the things that confused me when I first started working on the codebase. the "result function" was sometimes referred to as "the combiner" (which I think in terms of semantics makes the most sense). But it was occasionally referred to as the "output selector", but the genereated selector was also sometimes referred to as the "output selector". In the JSDocs I always called the final generated selector "a memoized output selector". 

BTW I did a quick search in github on `ParametricSelector` and `OutputParametricSelector` seems like anybody was hardly using them so I think we'll be okay removing them, I will go ahead and document their removal.

I still think in terms of simplicity "output selector" makes sense. The "memoized" part is implied and kind of a given. Input -> result function -> output
 sums up the flow of code as well. It's self explanatory in that Reselect goes through the "inputs" then "resultFunc" and gives an "output". It's linear and easier to digest no?

- <a name="dependencies"></a>[**`Dependencies`**](#dependencies): Same as [`input selectors`]. They are what the [`output selector`] "depends" on.

The below example serves as a visual aid:

```ts
const outputSelector = createSelector(
  [inputSelector1, inputSelector2, inputSelector3], // synonymous with `dependencies`.
  combiner // synonymous with `Result Function` or `resultFunc`.
)
```

</details>

<div align="right">[ <a href="#table-of-contents">↑ Back to top ↑</a> ]</div>

---

## How Does `Reselect` Work?

<a id="cascadingmemoization"></a>

### Cascading Memoization

> <details><summary><b>Click to expand</b></summary>

Reviewer: :hand: couple thoughts here:

I'd recently added this snippet under the "how do you pass in arguments?" section:

```ts
const finalSelector = (...args) => {
  const extractedValues = inputFunctions.map(input => input(...args));
  return output(...extractedValues);
}
```

I'd like to have that right under the `## How Does Reselect Work?` header, and one or two short paragraphs of explanation.  

Also, I'd like to avoid having headers with _no_ content at all. Let's have at least a sentence or two under each header with a basic explanation, then use the `<details>` to let them dig in further.

Author: > I'd recently added this snippet under the "how do you pass in arguments?" section:

Honestly the snippet confused me, but I think I know what you're trying to convey. And I already had an explanation of what you're trying to get across with the snippet, this is what it looks like:

```ts
const selectAvailableItems = createSelector(
  [
    // First input selector extracts items from the state
    (state: RootState) => state.items,
    // Second input selector forwards the category argument
    (state: RootState, category: string) => category,
    // Third input selector forwards the ID argument
    (state: RootState, category: string, id: number) => id
  ],
  // Output selector uses the extracted items, category, and ID
  (items, category, id) =>
    items.filter(item => item.category === category && item.id !== id)
)
```

Internally Reselect is doing this:

```ts
// Input selector #1
const items = (state: RootState, category: string, id: number) => state.items
// Input selector #2
const category = (state: RootState, category: string, id: number) => category
// Input selector #3
const id = (state: RootState, category: string, id: number) => id
// result of `output selector`
const finalResult =
  // The `Result Function`
  items.filter(item => item.category === category && item.id !== id)
```

> Also, I'd like to avoid having headers with no content at all. Let's have at least a sentence or two under each header with a basic explanation, then use the <details> to let them dig in further.

That makes a lot of sense, I will fix it. :)

Reviewer: Yeah, my example is meant to show that "_all_ arguments get passed to _all_ input selectors", specifically (which is why they all have to have compatible signatures).

Author: This is how I explained it:

1. **Consistency in Arguments**: Ensure that all positional arguments across [input selectors] are of the same type for consistency.

2. **Selective Argument Usage**: Design each selector to use only its relevant argument(s) and ignore the rest. This is crucial because all [input selectors] receive the same arguments that are passed to the [output selector].

Do you still want me to include your snippet as well?

Reviewer: Yeah, I've used variations of that snippet a few times and definitely want it in here.

The way `Reselect` works can be broken down into multiple parts:

> 1. **Initial Run**: On the first call, `Reselect` runs all the [`input selectors`], gathers their results, and passes them to the [`result function`].

Author: Just wanted to make jumping to definitions easier, kind of like how VSCode works with TS and type definitions.

2. **Subsequent Runs**: For subsequent calls, `Reselect` performs two levels of checks:

   - **First Level**: It compares the current arguments with the previous ones (done by `argsMemoize`).

     - If they're the same, it returns the cached result without running the [`input selectors`] or the [`result function`].

     - If they differ, it proceeds to the second level.

   - **Second Level**: It runs the [`input selectors`] and compares their current results with the previous ones (done by `memoize`).
     > [!NOTE]
     > If any one of the [`input selectors`] return a different result, all [`input selectors`] will recalculate.
     - If the results are the same, it returns the cached result without running the [`result function`].
     - If the results differ, it runs the [`result function`].

> This behavior is what we call **_Cascading Double-Layer Memoization_**.

Author: If `argsMemoize` fails, Reselect "cascades" to `memoize`. I will tweak the explanation to make it more clear.

store.dispatch(toggleRead(0))
selectCompletedTodos(store.getState()) // The `input selectors` will run, But the `result function` is skipped and the cached result will be returned.
```

Even when the overall `state` changes, `Reselect` ensures efficient memoization through its unique approach. The [`result function`] doesn't re-run if the relevant part of the `state` (in this case `state.todos`), remains unchanged. This is due to `Reselect`'s [**_Cascading Double-Layer Memoization_**][**_Cascading Memoization_**]. The first layer checks the entire `state`, and the second layer checks the results of the [`input selectors`]. If the first layer fails (due to a change in the overall `state`) but the second layer succeeds (because `state.todos` is unchanged), `Reselect` skips recalculating the [`result function`]. This dual-check mechanism makes `Reselect` particularly effective in `Redux` applications, ensuring computations are only done when truly necessary.

</details>

<div align="right">[ <a href="#table-of-contents">↑ Back to top ↑</a> ]</div>

---

## API

<a id="createselector"></a>

### createSelector(...inputSelectors | [inputSelectors], resultFunc, createSelectorOptions?)

> <details><summary><b>Description</b></summary>

Reviewer: :hand: "Description", "Parameters", and "Returns" shouldn't be hidden inside of `<details>` - we want readers to be able to see those directly as they scroll down.

Author: Yessirrr. I will adjust that.

Accepts one or more ["input selectors"][`input selectors`] (either as separate arguments or a single array),
a single ["result function"][`result function`] / ["combiner"][`combiner`], and an optional options object, and
generates a memoized selector function.

</details>

<details><summary><b>Parameters</b></summary>

| Name                     | Description                                                                                      |
| :----------------------- | :----------------------------------------------------------------------------------------------- |

> | `inputSelectors`         | An array of [`input selectors`], can also be passed as separate arguments.                       |

Reviewer: :hand: We'll to update the terms consistently, but especially here.

Author: What "term" are we talking about here?

Reviewer: Good question, what _was_ I thinking? :)

It was one of two things:

- We can skip wrapping every single use of one of the "Terminology" terms in links.  Per one of my other comments, we only need to highlight those in the first usage (and we only want to use backticks for actual variable/function names)
- Making sure that whatever final terminology terms we settle on get used consistently, particularly in the API reference section

Author: Yes sir, I will fix the backticks and formatting issues.

| `Result`                      | The return type of the [`combiner`] as well as the [`output selector`].                                                                                                                                |
| `OverrideMemoizeFunction`     | The type of the optional `memoize` function that could be passed into the options object to override the original `memoize` function that was initially passed into [`createSelectorCreator`].         |
| `OverrideArgsMemoizeFunction` | The type of the optional `argsMemoize` function that could be passed into the options object to override the original `argsMemoize` function that was initially passed into [`createSelectorCreator`]. |

</details>

<div align="right">[ <a href="#table-of-contents">↑ Back to top ↑</a> ]</div>

---

### Memoization Functions

<a id="defaultmemoize"></a>

#### defaultMemoize(func, equalityCheckOrOptions = defaultEqualityCheck)

<details><summary><b>Description</b></summary>

The standard memoize function used by [`createSelector`].

> It has a default cache size of 1. This means it always recalculates when the value of an argument changes. However, this can be customized as needed with a specific max cache size (**`Since`** 4.1.0).

Reviewer: Don't need to format "since" anywhere:

```suggestion
It has a default cache size of 1. This means it always recalculates when the value of an argument changes. However, this can be customized as needed with a specific max cache size (since 4.1.0).
```

Author: Initially got that from [markdown-magic](#https://github.com/DavidWells/markdown-magic). But you're right I will fix it :)

It determines if an argument has changed by calling the `equalityCheck` function. As `defaultMemoize` is designed to be used with immutable data, the default `equalityCheck` function checks for changes using [`reference equality`][Reference Equality Check]:

```ts
const defaultEqualityCheck = (previousValue: any, currentValue: any) => {
  return previousValue === currentValue
}
```

</details>

<details><summary><b>Parameters</b></summary>

| Name                     | Description                                                 |
| :----------------------- | :---------------------------------------------------------- |
| `func`                   | The function to be memoized.                                |
| `equalityCheckOrOptions` | Either an `equality check` function or an `options` object. |

> **`Since`** 4.1.0, `defaultMemoize` also accepts an options object as its first argument instead of an `equalityCheck` function. The `options` object may contain:

Reviewer: ```suggestion
Since 4.1.0, `defaultMemoize` also accepts an options object as its first argument instead of an `equalityCheck` function. The `options` object may contain:
```

| `resultEqualityCheck` | If provided, used to compare a newly generated output value against previous values in the cache. If a match is found, the old value is returned. This addresses the common <code>todos.map(todo => todo.id)</code> use case, where an update to another field in the original data causes a recalculation due to changed references, but the output is still effectively the same. |
| `maxSize`             | The cache size for the selector. If greater than 1, the selector will use an LRU cache internally. <br /> **`Default`** = 1                                                                                                                                                                                                                                                         |

> > [!WARNING]

Reviewer: :hand: This "warning" block doesn't seem to be styled right - should be this, I think:

```suggestion
> **Warning**
```

Author: They're the new [alert badges](#https://github.com/orgs/community/discussions/16925) in github-flavored markdown

Reviewer: Huh, okay, hadn't seen that variation.

Author: Kinda neat isn't it?

> If `resultEqualityCheck` is used inside `argsMemoizeOptions` it has no effect.

</details>

<details><summary><b>Returns</b></summary>

A memoized function with a `.clearCache()` method attached.

</details>

<details><summary><b>Type parameters</b></summary>

| Name   | Description                                |
| :----- | :----------------------------------------- |
| `Func` | The type of the function that is memoized. |

</details>

> <details><summary><b><code>Examples</code></b></summary>

Reviewer: :hand: don't need to format "Examples":

```suggestion
<details><summary><b>Examples</b></summary>
```

Author: yesssirrr

> ###### Using [`createSelector`]

Reviewer: :hand: Can we tweak these two titles?  Maybe emphasize "Using `defaultMemoize` with `createSelector`", etc.

Author: Gladddddllllyyyy :)

    how long values are kept in cache as it's based on garbage collection and [`WeakMap`]s.
- Cons:
  - There's currently no way to alter the argument comparisons. They're based on strict reference equality.

</details>

<details><summary><b>Use Cases</b></summary>

- This memoizer is likely best used for cases where you need to call the
  same selector instance with many different arguments, such as a single
  selector instance that is used in a list item component and called with
  item IDs like:

```ts
useSelector(state => selectSomeData(state, id))
```

</details>

<details><summary><b>Parameters</b></summary>

| Name   | Description                  |
| :----- | :--------------------------- |
| `func` | The function to be memoized. |

</details>

<details><summary><b>Returns</b></summary>

A memoized function with a `.clearCache()` method attached.

</details>

<details><summary><b>Type parameters</b></summary>

| Name   | Description                                |
| :----- | :----------------------------------------- |
| `Func` | The type of the function that is memoized. |

</details>

<details><summary><b><code>Examples</code></b></summary>

> Prior to `weakMapMemoize`, you had this problem:

Reviewer: :hand: This section feels like a "use case comparison", rather than "examples".

    todos.filter(todo => todo.id === id)
  )

const MyComponent: FC<Props> = ({ id }) => {
  const selectTodosById = useMemo(() => parametricSelector(id), [id])

  const todosById = useSelector(selectTodosById)

  return (
    <div>
      {todosById.map(todo => (
        <div key={todo.id}>{todo.title}</div>
      ))}
    </div>
  )
}
```

3. Using [`useCallback`].

```tsx
const parametricSelector = createSelector(
  [(state: RootState) => state.todos, (state: RootState, id: number) => id],
  (todos, id) => todos.filter(todo => todo.id === id)
)

const MyComponent: FC<Props> = ({ id }) => {
  const selectTodosById = useCallback(parametricSelector, [])

  const todosById = useSelector(state => selectTodosById(state, id))

  return (
    <div>
      {todosById.map(todo => (
        <div key={todo.id}> {todo.title}</div>
      ))}
    </div>
  )
}
```

4. Use [`re-reselect`]:

```ts
import { createCachedSelector } from 're-reselect'

const parametricSelector = createCachedSelector(
  [(state: RootState) => state.todos, (state: RootState, id: number) => id],
  (todos, id) => todos.filter(todo => todo.id === id)
)((state: RootState, id: number) => id)
```

> Starting in 5.0.0, you can eliminate this problem using `weakMapMemoize`.

Reviewer: :hand: If we're going to have this kind of comparison, there ought to be more information here explaining _why/how_ `weakMapMemoize` solves this issue.

Author: I will the additional info. Just to be sure, we are specifically refering to the "cache size" problem right?

Reviewer: Yep!

###### Using [`createSelector`]

```ts

> const parametricSelector = createSelector(

Reviewer: :hand: tbh I've never liked the phrase "parametric selector", and I don't think using it as a name in this example helps.  

Also, doing a `.filter()` with _one_ ID in comparison seems unlikely :)  How about filtering by "category" instead?  Seems like something that you might want to do many times.

Author: I will update the example, do you want me to name it something generic, what would we call a selector that takes additional parameters? or perhaps we don't actually need to specify that and we can call it something generic "selectSomething"? to stay consistent with the "selectX" pattern?

Reviewer: Try to name them meaningfully based on the example (like `selectTodosByCategory`, etc), or just `selectSomething` as a literal fallback, yeah

Author: Yeah I will change all the examples to match the `selectX` pattern for consistency.

  [(state: RootState) => state.todos, (state: RootState, id: number) => id],
  (todos, id) => todos.filter(todo => todo.id === id),
  {
    memoize: weakMapMemoize,
    argsMemoize: weakMapMemoize
  }
)

parametricSelector(state, 0) // Selector runs
parametricSelector(state, 0)
parametricSelector(state, 1) // Selector runs
parametricSelector(state, 0)
```

###### Using [`createSelectorCreator`]

```ts
import { createSelectorCreator, weakMapMemoize } from 'reselect'

const createSelectorWeakMap = createSelectorCreator({
  memoize: weakMapMemoize,
  argsMemoize: weakMapMemoize
})

const parametricSelector = createSelector(
  [(state: RootState) => state.todos, (state: RootState, id: number) => id],
  (todos, id) => todos.filter(todo => todo.id === id)
)

parametricSelector(state, 0) // Selector runs
parametricSelector(state, 0)
parametricSelector(state, 1) // Selector runs
parametricSelector(state, 0)
```

This solves the problem of having to know and set the cache size prior to creating a memoized selector.

</details>

<div align="right">[ <a href="#table-of-contents">↑ Back to top ↑</a> ]</div>

---

<a id="autotrackmemoize"></a>

> #### autotrackMemoize(func) - (**`Since`** 5.0.0)

Reviewer: :hand: Let's 100% refer to this as `unstable_autotrackMemoize` everywhere in the docs until it's actually ready for release.

Author: Surething

<details><summary><b>Description</b></summary>

> Uses an "auto-tracking" approach inspired by the work of the Ember Glimmer team. It uses a Proxy to wrap arguments and track accesses to nested fields in your selector on first read. Later, when the selector is called with new arguments, it identifies which accessed fields have changed and only recalculates the result if one or more of those accessed fields have changed. This allows it to be more precise than the shallow equality checks in defaultMemoize.

Reviewer: :hand: Add a warning under this section clarifying that it is still experimental and undergoing testing

| :----- | :--------------------------- |
| `func` | The function to be memoized. |

</details>

<details><summary><b>Returns</b></summary>

A memoized function with a `.clearCache()` method attached.

</details>

<details><summary><b>Type parameters</b></summary>

| Name   | Description                                |
| :----- | :----------------------------------------- |
| `Func` | The type of the function that is memoized. |

</details>

<details><summary><b><code>Examples</code></b></summary>

###### Using [`createSelector`]

```ts
import {
  unstable_autotrackMemoize as autotrackMemoize,
  createSelector
} from 'reselect'

const selectTodoIds = createSelector(
  [(state: RootState) => state.todos],
  todos => todos.map(todo => todo.id),
  { memoize: autotrackMemoize }
)
```

###### Using [`createSelectorCreator`]

```ts
import {
  unstable_autotrackMemoize as autotrackMemoize,
  createSelectorCreator
} from 'reselect'

const createSelectorAutotrack = createSelectorCreator({
  memoize: autotrackMemoize
})

const selectTodoIds = createSelectorAutotrack(
  [(state: RootState) => state.todos],
  todos => todos.map(todo => todo.id)
)
```

> - Pros:

Reviewer: :hand: Looks like there's some copy-paste duplication from the description above? Can remove this.

Author: Yikes, I thought I got rid of that, guess not

  - It is likely to avoid excess calculations and recalculate fewer times than defaultMemoize will, which may also result in fewer component re-renders.
- Cons:

  - It only has a cache size of 1.
  - It is slower than defaultMemoize, because it has to do more work. (How much slower is dependent on the number of accessed fields in a selector, number of calls, frequency of input changes, etc)
  - It can have some unexpected behavior. Because it tracks nested field accesses, cases where you don't access a field will not recalculate properly. For example, a badly-written selector like:

  ```ts
  createSelector([state => state.todos], todos => todos)
  ```

  that just immediately returns the extracted value will never update, because it doesn't see any field accesses to check.

**Use Cases for `autotrackMemoize`:**

- It is likely best used for cases where you need to access specific nested fields in data, and avoid recalculating if other fields in the same data objects are immutably updated.

</details> -->

<div align="right">[ <a href="#table-of-contents">↑ Back to top ↑</a> ]</div>

---

<a id="createselectorcreator" ></a>

> ### createSelectorCreator(memoize | options, ...memoizeOptions)

Reviewer: :hand: let's move this up to be directly after `createSelector`.  I think the reasonable ordering is:

- `createSelector`
- `createSelectorCreator`
- `createStructuredSelector`
- memoization functions

Author: It was ordered that way initially but you're correct, I will rearrange them.

import isEqual from 'lodash.isequal'

// create a "selector creator" that uses lodash.isequal instead of ===
const createDeepEqualSelector = createSelectorCreator(defaultMemoize, isEqual)

// use the new "selector creator" to create a selector
const selectSum = createDeepEqualSelector(
  [state => state.values.filter(val => val < 5)],
  values => values.reduce((acc, val) => acc + val, 0)
)
```

###### Use memoize function from `Lodash` for an unbounded cache

```js
import { createSelectorCreator } from 'reselect'
import memoize from 'lodash.memoize'

const hashFn = (...args) =>
  args.reduce((acc, val) => acc + '-' + JSON.stringify(val), '')

const customSelectorCreator = createSelectorCreator(memoize, hashFn)

const selector = customSelectorCreator(
  [state => state.a, state => state.b],
  (a, b) => a + b
)
```

</details>

<div align="right">[ <a href="#table-of-contents">↑ Back to top ↑</a> ]</div>

---

<a id="createstructuredselector"></a>

### createStructuredSelector({ inputSelectors }, selectorCreator = createSelector)

<details><summary><b>Description</b></summary>

A convenience function for a common pattern that arises when using `Reselect`.

> The selector passed to a `connect` decorator often just takes the values of its [`input selectors`]

Reviewer: :hand: let's ditch the `connect` reference, and instead say it simplifies returning an object made up of selector results.

Author: I could not agree more. It seems like `createStructuredSelector` was created specifically to be used with `connect` and now that `connect` is legacy API, it seems like `createStructuredSelector` is also kind of a legacy API.

Reviewer: Mostly, yeah.  You _could_ use it with `useSelector` to do `const {a, b} = useSelector(someStructuredSelector)`, and that's still a sort of plausible use case.  (Or you could just do `const {a, b} = useSelector(state => ({a: state.a, b: state.b}), shallowEqual), and have basically the same result.

```ts
state => ({ a: state.a, b: state.b })
```

or

```ts
state => state.todos.map(todo => todo.id)
```

that will cause the selector to never memoize properly.
Since this is a common mistake, we've added a development mode check to catch this. By default, [`createSelector`] will now run the [`input selectors`] twice during the first call to the selector. If the result appears to be different for the same call, it will log a warning with the arguments and the two different sets of extracted input values.

```ts
type StabilityCheckFrequency = 'always' | 'once' | 'never'
```

| Possible Values | Description                                     |
| :-------------- | :---------------------------------------------- |
| `once`          | Run only the first time the selector is called. |
| `always`        | Run every time the selector is called.          |
| `never`         | Never run the `input stability check`.          |

> [!IMPORTANT]

> > The `input stability check` is automatically disabled in production environments.

Reviewer: :hand: don't need to format words like these.

```suggestion
> The input stability check is automatically disabled in production environments.
```

Author: Yes sire!

</details>

<a id="outputselectorfields"></a>

### Output Selector Fields

<details><summary><b>The <code>output selectors</code> created by <code>createSelector</code> have several additional properties attached to them:</b></summary>

| Name                            | Description                                                                                                                                                                                       |
| ------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `resultFunc`                    | The final function passed to [`createSelector`]. Otherwise known as the `combiner`.                                                                                                               |
| `memoizedResultFunc`            | The memoized version of `resultFunc`.                                                                                                                                                             |

> | `lastResult`                    | Returns The last result calculated by `memoizedResultFunc`.                                                                                                                                       |

Reviewer: ```suggestion
| `lastResult`                    | Returns the last result calculated by `memoizedResultFunc`.                                                                                                                                       |
```

Author: You have the eyes of an eagle!

| `resetDependencyRecomputations` | Resets the `dependencyRecomputations` count to 0.                                                                                                                                                 |
| `memoize`                       | Function used to memoize the `resultFunc`.                                                                                                                                                        |
| `argsMemoize`                   | Function used to memoize the arguments passed into the [`output selector`].                                                                                                                       |

</details>

<div align="right">[ <a href="#table-of-contents">↑ Back to top ↑</a> ]</div>

---

<a id="v5summary"></a>

## What's New in 5.0.0?

<details><summary><b>Version 5.0.0 introduces several new features and improvements:</b></summary>

- **Customization Enhancements**:

  - Added the ability to pass an options object to [`createSelectorCreator`], allowing for customized `memoize` and `argsMemoize` functions, alongside their respective options (`memoizeOptions` and `argsMemoizeOptions`).

>   - The [`createSelector`] function now supports direct customization of `memoize` and `argsMemoize` within its options object.

Reviewer: :hand: I'd put this one above the `createSelectorCreator` customization, really :)

Author: You wish is my command :)

---

## Additional Tips & Tricks

<details><summary><b>Top Level Selectors Pattern</b></summary>

This pattern simplifies the creation of selectors in your application. Start by defining a generic type for your top-level selectors:

```ts
import type { Selector } from 'reselect'

export type TopLevelSelectors<State> = {
  [K in keyof State as K extends string
    ? `select${Capitalize<K>}`
    : never]: Selector<State, State[K], never>
}
```

With this setup, you can easily create top-level selectors for your application's state:

```ts
const topLevelSelectors: TopLevelSelectors<RootState> = {
  selectAlerts: state => state.alerts,
  selectTodos: state => state.todos,
  selectUsers: state => state.users
}
```

This approach allows for more streamlined and readable selector creation. For example:

```ts
const selectCompletedTodos = createSelector(
  [topLevelSelectors.selectTodos],
  todos => todos.filter(todo => todo.completed === true)
)
```

</details>

<div align="right">[ <a href="#table-of-contents">↑ Back to top ↑</a> ]</div>

---

## FAQ

### Why isn’t my selector recomputing when the input state changes?

> <details><summary><b>Answer</b></summary>

Reviewer: :hand: Actually going to remove most of these `<details>` sections from the FAQs, except for maybe a couple longer answers
